modules/drivers/redshift/src/metabase/driver/redshift.clj

```diff
@@ -251,3 +253,27 @@
                               :filter_values       (field->parameter-value query)})
        " */ "
        (qputil/default-query->remark query)))
+
+(defn- make-schema-usage-filter-xf [conn]
+  (fn [xf]
+    (let [stmt (prepare-statement conn "SELECT HAS_SCHEMA_PRIVILEGE(?, 'USAGE');")]
+      (fn
+        ([]
+         (xf))
+        ([result]
+         (.close stmt)
+         (xf result))
+        ([result table-schema]
+         (with-open [rs (.executeQuery (doto stmt (.setString 1 table-schema)))]
+           (let [has-perm? (and (.next rs)
+                                (.getBoolean rs 1))]
+             (if has-perm?
+               (xf result table-schema)
+               (do (log/tracef "Ignoring schema %s because no USAGE privilege on it" table-schema)
+                   result)))))))))
```

Comment:
If any steps here throw an Exception it's going to leave the `PreparedStatement` open without closing it.

Sorry to keep making your life hard, but I think what you actually want to do here is reify `clojure.lang.IReduceInit` so you can use `with-open` with `prepare-statement`. Give me a second and I can show you what I mean